---
title: "HEK cells"
output:
  html_notebook:
    code_folding: hide
    css: ~/Documents/common/utils/Rnotebook_style.css
    theme: united
    toc: yes
    toc_float: no
    urlcolor: blue
  html_document:
    code_folding: hide
    css: ~/Documents/common/utils/Rnotebook_style.css
    theme: united
    toc: yes
    toc_float: no
    urlcolor: blue
---
```{r global_options, echo = FALSE, include = FALSE}
knitr::opts_chunk$set(echo = T, warning = FALSE, message = FALSE,
                      cache = FALSE, tidy = TRUE, size = "small")
```

```{r}
source("~/Documents/common/utils/dataIO.R")
source("~/Documents/common/utils/tophelpers.R")
source("~/Documents/common/utils/common_utils.R")

parse_stats <- function(df){
  df %>%
      melt(id = 1:2, variable.name = "metric") %>%
      filter(!grepl("star.*on$", metric)) %>%
      mutate(metric = as.character(metric),
             value = gsub("%", "", value),
             value = as.numeric(value),
             value = ifelse(metric %in% c("riboPct", "mitoPct", "rRNA45sPct"), value/100, value),
             exp = ifelse(expId == "EXPsmart3", "Benchmark", expId),
             exp = ifelse(expId == "EXP_HEK_benchmark", "Benchmark", expId),
             exp = gsub("_HEK_bead_notrim|_P01_HEK", "", exp),
             metric = ifelse(grepl("_", metric), metric, paste0("fastq_", metric)),
             metric = gsub("picard", "picard::", metric)) %>%
      separate(metric, c("source", "metric"), sep = "_", extra = "merge") %>%
      filter(source != "biotype", !is.na(value))
}

df_metrics <- read_tsv("HEK_data/HEK_stats_complete.tsv") %>% parse_stats()
df_metrics_nogel <- read_tsv("HEK_nogel_data/HEK_stats_complete.tsv") %>% parse_stats() %>%
  mutate(exp = gsub("EXP0599_rawpool_rerun", "EXP0599_pre_gel_rerun", exp),
         exp = gsub("EXPsmart3_HEK", "Benchmark_newMito", exp))

df_metrics %<>% bind_rows(df_metrics_nogel)

df_key <- df_metrics %>% select(metric) %>% unique()

```



# General reads alignment reads stat
```{r fig.width=5, fig.height=1.5}
targeted_metrics <- c("Numberofinputreads", "totalmappedreads%")

df_metrics %>%
  filter(metric %in% targeted_metrics) %>%
  mutate(metric = gsub("of", " of ", metric),
         metric = gsub("reads", " reads ", metric),
         metric = gsub("mapped", " mapped ", metric),
         metric = gsub("un mapped", "unmapped ", metric)) %>%
  ggplot(aes(exp, value)) +
  # geom_boxplot(aes(color = expId)) +
  geom_jitter(aes(color = exp), width = 0.2, shape = 1, size = 2) +
  facet_wrap(~metric, scales = "free_y") +
  theme_bw() + 
  theme(axis.text.x = element_blank()) 
```

```{r fig.width=8, fig.height=2}
targeted_metrics <- c("Uniquelymappedreads%", "%ofreadsmappedtomultipleloci", "%ofreadsmappedtotoomanyloci",
                      "%ofreadsunmappedtoomanymismatches", "%ofreadsunmappedtooshort", "%ofreadsunmappedother")

df_metrics %>%
  filter(metric %in% targeted_metrics) %>%
  mutate(metric = gsub("of", " of ", metric),
         metric = gsub("reads", " reads ", metric),
         metric = gsub("mapped", "mapped ", metric),
         metric = gsub("toomany", " too many ", metric),
         metric = gsub("multiple", " multiple ", metric)) %>%
  ggplot(aes(cellId, value, group = cellId)) +
  geom_bar(stat = "identity", aes(fill = metric), position='stack') +
  scale_fill_brewer(palette = "Set2") +
  facet_wrap(~exp, scales = "free_x", nrow = 1) +
  theme_bw() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) 
```



```{r fig.width=5, fig.height=2}
df_metrics %>%
  filter(grepl("RIME_BIAS", metric)) %>%
  mutate(metric = gsub("of", " of ", metric),
         metric = gsub("reads", " reads ", metric)) %>%
  ggplot(aes(exp, value)) +
  # geom_boxplot(aes(color = expId)) +
  geom_jitter(aes(color = exp), width = 0.2, shape = 1, size = 2) +
  facet_wrap(~metric, scales = "free_y") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 

```

# rRNA and mito content
```{r fig.width=5, fig.height=2}
targeted_metrics <- c("riboPct", "mitoPct", "rRNA45sPct")
df_metrics %>%
  filter(metric %in% targeted_metrics) %>%
  ggplot(aes(exp, value)) +
  # geom_boxplot(aes(color = expId)) +
  geom_jitter(aes(color = expId), width = 0.2, shape = 1, size = 2) +
  facet_wrap(~metric, scales = "free_y") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```
```{r fig.width=6, fig.height=2}
targeted_metrics <- c("mitoPct", "mtRNR1count", "mtRNR2count", "mitoHumaninPct")
df_metrics %>%
  filter(metric %in% targeted_metrics) %>%
  ggplot(aes(exp, value)) +
  # geom_boxplot(aes(color = expId)) +
  geom_jitter(aes(color = expId), width = 0.2, shape = 1, size = 2) +
  facet_wrap(~metric, scales = "free_y", nrow = 1) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

# Alignment on gene body
```{r fig.width=5, fig.height=2}
targeted_metrics <- c("PCT_RIBOSOMAL_BASES", "PCT_CODING_BASES", "PCT_UTR_BASES", "PCT_INTRONIC_BASES", "PCT_INTERGENIC_BASES")

df_metrics %>%
  filter(metric %in% targeted_metrics) %>%
  ggplot(aes(exp, value)) +
  geom_boxplot(aes(color = expId)) +
  # geom_jitter(aes(color = expId), width = 0.2, shape = 1, size = 2) +
  facet_wrap(~metric , scales = "free_y", nrow = 2) +
  theme_bw()
```
```{r fig.width=6, fig.height=2}
targeted_metrics <- c("PCT_RIBOSOMAL_BASES", 
                      "PCT_CODING_BASES", "PCT_UTR_BASES", "PCT_INTRONIC_BASES",
                      "PCT_INTERGENIC_BASES")

df_metrics %>%
  filter(metric %in% targeted_metrics) %>%
  mutate(metric = gsub("PCT_", "% ", metric),
         metric = gsub("_BASES", "", metric)) %>%
  ggplot(aes(cellId, value, group = cellId)) +
  geom_bar(stat = "identity", aes(fill = metric), position='stack') +
  scale_fill_brewer(palette = "Set2") +
  facet_wrap(~exp, scales = "free_x", nrow = 1) +
  theme_bw() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  ggtitle("Distribution of aligned bases per cell")
```
# Top features correlated with number of genes detected

```{r}
mat <- Seurat::Read10X_h5("HEK_data//rsem_count_matrix.h5")
df_gene_anno <- read_csv("~/Documents/common/resource//genome_anno/gene_symbol_GRCh38.csv") %>%
  mutate(gene_id = ensembl_gene_id) %>%
  select(gene_id, gene_symbol, gene_biotype, gene_gc_pct) %>%
  filter(gene_biotype == "protein_coding")

df_gene_detected <- data.frame(
  cellId = colnames(mat),
  num_genes = Matrix::colSums(mat > 0),
  num_pt_genes = Matrix::colSums(mat[rownames(mat) %in% df_gene_anno$gene_symbol, ] > 0)
)

df_compare <- df_gene_detected %>%
  filter(num_genes > 1000) %>%
  left_join(df_metrics) %>%
  filter(source != "rsem")

df_compare_corr <- df_compare %>%
  group_by(source, metric) %>%
  mutate(stdev = sd(value)) %>%
  filter(stdev > 0) %>%
  summarise(pearson = cor(num_genes, value, method = "pearson", use = "complete.obs"),
            spearman = cor(num_genes, value, method = "spearman", use = "complete.obs")) %>%
  ungroup() %>%
  mutate(pe_rank = rank(-pearson, ties.method = "min"),
         sp_rank = rank(-spearman, ties.method = "min"),
         rank = pe_rank + sp_rank)
```
```{r fig.width=6, fig.height=3.5}
ggplot(df_compare_corr, aes(pearson, spearman)) +
  geom_point(aes(color = source)) +
  ggrepel::geom_text_repel(data = df_compare_corr %>% filter(rank < 10),
                           aes(pearson, spearman, label = metric),
                           size = 3) +
  ggrepel::geom_text_repel(data = df_compare_corr %>% filter(pearson < -0.2, spearman < -0.2),
                           aes(pearson, spearman, label = metric),
                           size = 3) +
  xlim(-0.6, 0.9) +
  ylim(-0.6, 0.9) +
  theme_bw() +
  ggtitle("Correlation with Number of genes detected")
```
```{r}
DT::datatable(df_compare_corr %>% filter(rank < 25 | rank > 160) %>%
                mutate(pearson = round(pearson, 3),
                       spearman = round(spearman, 3)) %>%
                arrange(-pearson),
              options = list(pageLength = 10))
```


```{r fig.width=6, fig.height=2.5}
targeted_metrics1 <- c("MEDIAN_INSERT_SIZE", "MEDIAN_5PRIME_TO_3PRIME_BIAS", "CODING_BASES", "NumberofsplicesTotal")
targeted_metrics2 <- c("GC_NC_20_39", "PCT_INTERGENIC_BASES", 
                      "MEDIAN_CV_COVERAGE", "PCT_R1_TRANSCRIPT_STRAND_READS")
g1 <- df_compare %>%
  filter(metric %in% targeted_metrics1) %>%
  ggplot(aes(value, num_genes)) +
  geom_point(aes(color = exp)) +
  stat_smooth(method = "lm") +
  facet_wrap(~metric, scales = "free_x", nrow = 1) +
  theme_bw() 

g2 <- df_compare %>%
  filter(metric %in% targeted_metrics2) %>%
  ggplot(aes(value, num_genes)) +
  geom_point(aes(color = exp)) +
  stat_smooth(method = "lm") +
  facet_wrap(~metric, scales = "free_x", nrow = 1) +
  theme_bw() 

grid.arrange(g1, g2, nrow = 2)
```

# Top features correlated with number of protein-coding genes detected

```{r}
df_compare_pt_corr <- df_compare %>%
  group_by(source, metric) %>%
  mutate(stdev = sd(value)) %>%
  filter(stdev > 0) %>%
  summarise(pearson = cor(num_pt_genes, value, method = "pearson", use = "complete.obs"),
            spearman = cor(num_pt_genes, value, method = "spearman", use = "complete.obs")) %>%
  ungroup() %>%
  mutate(pe_rank = rank(-pearson, ties.method = "min"),
         sp_rank = rank(-spearman, ties.method = "min"),
         rank = pe_rank + sp_rank)
```
```{r fig.width=6, fig.height=3.5}
ggplot(df_compare_pt_corr, aes(pearson, spearman)) +
  geom_point(aes(color = source)) +
  ggrepel::geom_text_repel(data = df_compare_pt_corr %>% filter(rank < 15),
                           aes(pearson, spearman, label = metric),
                           size = 3) +
  ggrepel::geom_text_repel(data = df_compare_pt_corr %>% filter(pearson < -0.3, spearman < -0.3),
                           aes(pearson, spearman, label = metric),
                           size = 3) +
  xlim(-0.6, 0.9) +
  ylim(-0.6, 0.9) +
  theme_bw() +
  ggtitle("Correlation with number of protein-coding genes detected")
```

```{r}
DT::datatable(df_compare_pt_corr %>% filter(rank < 25 | rank > 160) %>%
                mutate(pearson = round(pearson, 3),
                       spearman = round(spearman, 3)) %>%
                arrange(-pearson),
              options = list(pageLength = 10))
```

```{r fig.width=6, fig.height=2.5}
targeted_metrics1 <- c("NumberofsplicesTotal", "MEDIAN_5PRIME_TO_3PRIME_BIAS", 
                      "CODING_BASES", "MEDIAN_INSERT_SIZE")
targeted_metrics2 <- c("PCT_RIBOSOMAL_BASES", "GC_NC_20_39", "PCT_INTERGENIC_BASES", "MEDIAN_CV_COVERAGE")

g1 <- df_compare %>%
  filter(metric %in% targeted_metrics1) %>%
  ggplot(aes(value, num_pt_genes)) +
  geom_point(aes(color = exp)) +
  stat_smooth(method = "lm") +
  facet_wrap(~metric, scales = "free_x", nrow = 1) +
  theme_bw() +
  theme(strip.text.x = element_text(size = 11))

g2 <- df_compare %>%
  filter(metric %in% targeted_metrics2) %>%
  ggplot(aes(value, num_pt_genes)) +
  geom_point(aes(color = exp)) +
  stat_smooth(method = "lm") +
  facet_wrap(~metric, scales = "free_x", nrow = 1) +
  theme_bw() +
  theme(strip.text.x = element_text(size = 11))


grid.arrange(g1, g2, nrow = 2)
```
# Metric heatmap
```{r}
mat_metrics <- df_metrics %>%
  group_by(metric) %>%
  summarise(median = median(value),
         mad = mad(value)) %>%
  filter(mad > 0) %>%
  mutate(cv = mad/median) %>%
  filter(cv > 0.1)

df_metrics_meta <- df_metrics %>%
  filter(metric %in% mat_metrics$metric) %>%
  select(metric, source) %>%
  unique() %>%
  set_rownames(NULL) %>%
  tibble::column_to_rownames("metric")

df_cell_meta <- df_compare %>%
  select(cellId, exp, num_genes, num_pt_genes) %>%
  unique() %>%
  arrange(num_pt_genes) %>%
  set_rownames(NULL) %>%
  tibble::column_to_rownames("cellId")
  
mat_metrics_norm <- df_metrics %>%
  filter(metric %in% mat_metrics$metric) %>%
  group_by(metric) %>%
  mutate(norm = (value - quantile(value, 0.05))/(quantile(value, 0.95) - quantile(value, 0.05))) %>%
  ungroup() %>%
  mutate(z = scale(norm)[, 1],
         z = ifelse(abs(z) > 3, NA, z)) %>%
  acast(metric ~ cellId, value.var = "z")
```

```{r fig.width=8, fig.height=7}
pheatmap::pheatmap(mat_metrics_norm[, rownames(df_cell_meta)], 
                   show_colnames = F,
                   cluster_cols = F,
                   annotation_row = df_metrics_meta,
                   annotation_col = df_cell_meta,
                   fontsize = 15,
                   fontsize_row = 10)
```

