{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b7bd057",
   "metadata": {},
   "outputs": [],
   "source": [
    "%config Completer.use_jedi = False\n",
    "\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "import json\n",
    "import pickle\n",
    "from collections import defaultdict, Counter, OrderedDict\n",
    "from datetime import datetime\n",
    "\n",
    "import h5py\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.model_selection import train_test_split\n",
    "from scipy.sparse import csc_matrix, csr_matrix\n",
    "\n",
    "from glmnet import LogitNet\n",
    "\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format = 'retina'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "99908c44",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "sc.settings.set_figure_params(dpi=120, facecolor='white')\n",
    "\n",
    "sys.path.append(\"/home/liwang/repo/scRNAseq/smartSeq/\")\n",
    "from lib.reference import get_gene_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "90b5c0e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_csc_matrix_h5(h5_file):\n",
    "    \"\"\"generic io function for reading sparse matrix h5\"\"\"\n",
    "\n",
    "    with h5py.File(h5_file, \"r\") as h5in:\n",
    "        matrix = csc_matrix((h5in[\"data\"], h5in[\"indices\"], h5in[\"indptr\"]), shape=h5in[\"shape\"])\n",
    "\n",
    "    return matrix\n",
    "\n",
    "def sum_sparse_matrix(matrix, axis=0):\n",
    "    \"\"\"Sum a sparse matrix along an axis.\"\"\"\n",
    "    axis_sum = np.asarray(matrix.sum(axis=axis))  # sum along given axis\n",
    "    max_dim = np.prod(axis_sum.shape)  # get the max dimension\n",
    "    return axis_sum.reshape((max_dim,))  # reshape accordingly\n",
    "\n",
    "\n",
    "def normalize_matrix(matrix, use_median=False):\n",
    "    \"\"\"depth and log normalize matrix\"\"\"\n",
    "    \n",
    "    counts_per_bc = sum_sparse_matrix(matrix, axis=0)\n",
    "    median_counts_per_bc = max(1.0, np.median(counts_per_bc))\n",
    "    print(median_counts_per_bc)\n",
    "    if use_median:\n",
    "        scaling_factors = median_counts_per_bc / np.clip(counts_per_bc, 1.0, None)\n",
    "    else:\n",
    "        scaling_factors = 1000 / np.clip(counts_per_bc, 1.0, None)\n",
    "\n",
    "    # Normalize each barcode's total count by median total count\n",
    "    m = matrix.copy().astype(np.float64)\n",
    "    sparsefuncs.inplace_column_scale(m, scaling_factors)\n",
    "\n",
    "    # Use log counts\n",
    "    m.data = np.log2(1 + m.data)\n",
    "    \n",
    "    return m\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00b83d71",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ec4563c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Barcode</th>\n",
       "      <th>lane</th>\n",
       "      <th>cell_bc</th>\n",
       "      <th>nCount_ADT</th>\n",
       "      <th>nFeature_ADT</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>celltype.l1</th>\n",
       "      <th>celltype.l2</th>\n",
       "      <th>celltype.l3</th>\n",
       "      <th>donor</th>\n",
       "      <th>time</th>\n",
       "      <th>Phase</th>\n",
       "      <th>umap1</th>\n",
       "      <th>umap2</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1_AAACCCAAGACATACA</td>\n",
       "      <td>L1</td>\n",
       "      <td>AAACCCAAGACATACA-9</td>\n",
       "      <td>5949</td>\n",
       "      <td>211</td>\n",
       "      <td>5864</td>\n",
       "      <td>1617</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 TCM</td>\n",
       "      <td>CD4 TCM_1</td>\n",
       "      <td>P1</td>\n",
       "      <td>7</td>\n",
       "      <td>G1</td>\n",
       "      <td>5.286920</td>\n",
       "      <td>5.635788</td>\n",
       "      <td>donor1234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1_AAACCCACAACTGGTT</td>\n",
       "      <td>L1</td>\n",
       "      <td>AAACCCACAACTGGTT-9</td>\n",
       "      <td>6547</td>\n",
       "      <td>217</td>\n",
       "      <td>5067</td>\n",
       "      <td>1381</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>P4</td>\n",
       "      <td>3</td>\n",
       "      <td>S</td>\n",
       "      <td>11.907538</td>\n",
       "      <td>-4.530682</td>\n",
       "      <td>donor1234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1_AAACCCACACGTACTA</td>\n",
       "      <td>L1</td>\n",
       "      <td>AAACCCACACGTACTA-9</td>\n",
       "      <td>3508</td>\n",
       "      <td>207</td>\n",
       "      <td>4786</td>\n",
       "      <td>1890</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK_2</td>\n",
       "      <td>P3</td>\n",
       "      <td>7</td>\n",
       "      <td>G1</td>\n",
       "      <td>2.371758</td>\n",
       "      <td>-8.360968</td>\n",
       "      <td>donor1234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1_AAACCCACAGCATACT</td>\n",
       "      <td>L1</td>\n",
       "      <td>AAACCCACAGCATACT-9</td>\n",
       "      <td>6318</td>\n",
       "      <td>219</td>\n",
       "      <td>6505</td>\n",
       "      <td>1621</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>P4</td>\n",
       "      <td>7</td>\n",
       "      <td>G1</td>\n",
       "      <td>12.371468</td>\n",
       "      <td>-5.079568</td>\n",
       "      <td>donor1234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1_AAACCCACATCAGTCA</td>\n",
       "      <td>L1</td>\n",
       "      <td>AAACCCACATCAGTCA-9</td>\n",
       "      <td>5195</td>\n",
       "      <td>213</td>\n",
       "      <td>4332</td>\n",
       "      <td>1633</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 TEM</td>\n",
       "      <td>CD8 TEM_1</td>\n",
       "      <td>P3</td>\n",
       "      <td>3</td>\n",
       "      <td>G1</td>\n",
       "      <td>5.819462</td>\n",
       "      <td>-3.657757</td>\n",
       "      <td>donor1234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Barcode lane             cell_bc  nCount_ADT  nFeature_ADT  \\\n",
       "0  L1_AAACCCAAGACATACA   L1  AAACCCAAGACATACA-9        5949           211   \n",
       "1  L1_AAACCCACAACTGGTT   L1  AAACCCACAACTGGTT-9        6547           217   \n",
       "2  L1_AAACCCACACGTACTA   L1  AAACCCACACGTACTA-9        3508           207   \n",
       "3  L1_AAACCCACAGCATACT   L1  AAACCCACAGCATACT-9        6318           219   \n",
       "4  L1_AAACCCACATCAGTCA   L1  AAACCCACATCAGTCA-9        5195           213   \n",
       "\n",
       "   nCount_RNA  nFeature_RNA celltype.l1 celltype.l2 celltype.l3 donor  time  \\\n",
       "0        5864          1617       CD4 T     CD4 TCM   CD4 TCM_1    P1     7   \n",
       "1        5067          1381       CD8 T   CD8 Naive   CD8 Naive    P4     3   \n",
       "2        4786          1890          NK          NK        NK_2    P3     7   \n",
       "3        6505          1621       CD8 T   CD8 Naive   CD8 Naive    P4     7   \n",
       "4        4332          1633       CD8 T     CD8 TEM   CD8 TEM_1    P3     3   \n",
       "\n",
       "  Phase      umap1     umap2      batch  \n",
       "0    G1   5.286920  5.635788  donor1234  \n",
       "1     S  11.907538 -4.530682  donor1234  \n",
       "2    G1   2.371758 -8.360968  donor1234  \n",
       "3    G1  12.371468 -5.079568  donor1234  \n",
       "4    G1   5.819462 -3.657757  donor1234  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cell_meta = pd.read_csv(\"data/tcell_metadata.csv\")\n",
    "df_cell_meta[\"batch\"] = df_cell_meta[\"donor\"].apply(lambda val: \"donor1234\" if val in [\"P1\", \"P2\", \"P3\", \"P4\"] else \"donor5678\")\n",
    "df_cell_meta.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "52dc9040",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# mat = sc.read_10x_h5(\"/home/liwang/yard/imPACT-SMARTseq/literature_data/seurat_v4_cite-seq/cellranger_reanalyze/seurat_v4_tcells_sample/outs/filtered_feature_bc_matrix.h5\", \n",
    "#                          gex_only=False)\n",
    "mat = sc.read_10x_h5(\"/home/liwang/yard/dataset/seurat_v4_cite-seq/cellranger_reanalyze/fbm.h5\",  gex_only=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b43e4ded",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL669831.5</th>\n",
       "      <td>AL669831.5</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00115</th>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM41C</th>\n",
       "      <td>ENSG00000230368</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_ids    feature_types  genome\n",
       "AL627309.1  ENSG00000238009  Gene Expression  GRCh38\n",
       "AL669831.5       AL669831.5  Gene Expression  GRCh38\n",
       "LINC00115   ENSG00000225880  Gene Expression  GRCh38\n",
       "FAM41C      ENSG00000230368  Gene Expression  GRCh38\n",
       "NOC2L       ENSG00000188976  Gene Expression  GRCh38"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mat.var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7f92d71",
   "metadata": {},
   "source": [
    "#### Select gene features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b8d23602",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((87918, 20957), (87628, 20177))"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# smart counter genes\n",
    "gene_anno = get_gene_anno()\n",
    "\n",
    "# seurat tcell genes\n",
    "seurat_tcell_genes = [gene.strip() for gene in open(\"data/CD4CD8_genes.txt\")]\n",
    "\n",
    "# cellranger genes\n",
    "cr_genes = mat.var_names[mat.var[\"feature_types\"] == \"Gene Expression\"]\n",
    "\n",
    "# common genes\n",
    "common_genes = set(gene_anno[\"name\"]).intersection(set(cr_genes))\n",
    "selected_genes = set(seurat_tcell_genes).intersection(common_genes)\n",
    "\n",
    "common_genes = sorted(list(common_genes))\n",
    "selected_genes = sorted(list(selected_genes))\n",
    "\n",
    "mat_gex = mat[df_cell_meta[\"cell_bc\"][df_cell_meta[\"celltype.l3\"] != \"CD8 Naive_2\"], mat.var[\"feature_types\"] == \"Gene Expression\"]\n",
    "mat_gex = mat_gex[:, common_genes]\n",
    "mat.shape, mat_gex.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "07552b8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Barcode</th>\n",
       "      <th>lane</th>\n",
       "      <th>nCount_ADT</th>\n",
       "      <th>nFeature_ADT</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>celltype.l1</th>\n",
       "      <th>celltype.l2</th>\n",
       "      <th>celltype.l3</th>\n",
       "      <th>donor</th>\n",
       "      <th>time</th>\n",
       "      <th>Phase</th>\n",
       "      <th>umap1</th>\n",
       "      <th>umap2</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGACATACA-9</th>\n",
       "      <td>L1_AAACCCAAGACATACA</td>\n",
       "      <td>L1</td>\n",
       "      <td>5949</td>\n",
       "      <td>211</td>\n",
       "      <td>5864</td>\n",
       "      <td>1617</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>CD4 TCM</td>\n",
       "      <td>CD4 TCM_1</td>\n",
       "      <td>P1</td>\n",
       "      <td>7</td>\n",
       "      <td>G1</td>\n",
       "      <td>5.286920</td>\n",
       "      <td>5.635788</td>\n",
       "      <td>donor1234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAACTGGTT-9</th>\n",
       "      <td>L1_AAACCCACAACTGGTT</td>\n",
       "      <td>L1</td>\n",
       "      <td>6547</td>\n",
       "      <td>217</td>\n",
       "      <td>5067</td>\n",
       "      <td>1381</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>P4</td>\n",
       "      <td>3</td>\n",
       "      <td>S</td>\n",
       "      <td>11.907538</td>\n",
       "      <td>-4.530682</td>\n",
       "      <td>donor1234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACACGTACTA-9</th>\n",
       "      <td>L1_AAACCCACACGTACTA</td>\n",
       "      <td>L1</td>\n",
       "      <td>3508</td>\n",
       "      <td>207</td>\n",
       "      <td>4786</td>\n",
       "      <td>1890</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK_2</td>\n",
       "      <td>P3</td>\n",
       "      <td>7</td>\n",
       "      <td>G1</td>\n",
       "      <td>2.371758</td>\n",
       "      <td>-8.360968</td>\n",
       "      <td>donor1234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAGCATACT-9</th>\n",
       "      <td>L1_AAACCCACAGCATACT</td>\n",
       "      <td>L1</td>\n",
       "      <td>6318</td>\n",
       "      <td>219</td>\n",
       "      <td>6505</td>\n",
       "      <td>1621</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>CD8 Naive</td>\n",
       "      <td>P4</td>\n",
       "      <td>7</td>\n",
       "      <td>G1</td>\n",
       "      <td>12.371468</td>\n",
       "      <td>-5.079568</td>\n",
       "      <td>donor1234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACATCAGTCA-9</th>\n",
       "      <td>L1_AAACCCACATCAGTCA</td>\n",
       "      <td>L1</td>\n",
       "      <td>5195</td>\n",
       "      <td>213</td>\n",
       "      <td>4332</td>\n",
       "      <td>1633</td>\n",
       "      <td>CD8 T</td>\n",
       "      <td>CD8 TEM</td>\n",
       "      <td>CD8 TEM_1</td>\n",
       "      <td>P3</td>\n",
       "      <td>3</td>\n",
       "      <td>G1</td>\n",
       "      <td>5.819462</td>\n",
       "      <td>-3.657757</td>\n",
       "      <td>donor1234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Barcode lane  nCount_ADT  nFeature_ADT  \\\n",
       "AAACCCAAGACATACA-9  L1_AAACCCAAGACATACA   L1        5949           211   \n",
       "AAACCCACAACTGGTT-9  L1_AAACCCACAACTGGTT   L1        6547           217   \n",
       "AAACCCACACGTACTA-9  L1_AAACCCACACGTACTA   L1        3508           207   \n",
       "AAACCCACAGCATACT-9  L1_AAACCCACAGCATACT   L1        6318           219   \n",
       "AAACCCACATCAGTCA-9  L1_AAACCCACATCAGTCA   L1        5195           213   \n",
       "\n",
       "                    nCount_RNA  nFeature_RNA celltype.l1 celltype.l2  \\\n",
       "AAACCCAAGACATACA-9        5864          1617       CD4 T     CD4 TCM   \n",
       "AAACCCACAACTGGTT-9        5067          1381       CD8 T   CD8 Naive   \n",
       "AAACCCACACGTACTA-9        4786          1890          NK          NK   \n",
       "AAACCCACAGCATACT-9        6505          1621       CD8 T   CD8 Naive   \n",
       "AAACCCACATCAGTCA-9        4332          1633       CD8 T     CD8 TEM   \n",
       "\n",
       "                   celltype.l3 donor  time Phase      umap1     umap2  \\\n",
       "AAACCCAAGACATACA-9   CD4 TCM_1    P1     7    G1   5.286920  5.635788   \n",
       "AAACCCACAACTGGTT-9   CD8 Naive    P4     3     S  11.907538 -4.530682   \n",
       "AAACCCACACGTACTA-9        NK_2    P3     7    G1   2.371758 -8.360968   \n",
       "AAACCCACAGCATACT-9   CD8 Naive    P4     7    G1  12.371468 -5.079568   \n",
       "AAACCCACATCAGTCA-9   CD8 TEM_1    P3     3    G1   5.819462 -3.657757   \n",
       "\n",
       "                        batch  \n",
       "AAACCCAAGACATACA-9  donor1234  \n",
       "AAACCCACAACTGGTT-9  donor1234  \n",
       "AAACCCACACGTACTA-9  donor1234  \n",
       "AAACCCACAGCATACT-9  donor1234  \n",
       "AAACCCACATCAGTCA-9  donor1234  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mat_gex.obs = mat_gex.obs.join(df_cell_meta.set_index(\"cell_bc\"))\n",
    "mat_gex.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "c6aa8f97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 20.6 s, sys: 7.15 s, total: 27.8 s\n",
      "Wall time: 27.8 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "((87918, 20957), (87628, 20177), (87628, 488))"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "sc.pp.normalize_total(mat_gex, target_sum=10000)\n",
    "sc.pp.log1p(mat_gex)\n",
    "sc.pp.scale(mat_gex, max_value=10)\n",
    "\n",
    "mat_gex_filtered = mat_gex[:, selected_genes]\n",
    "mat.shape, mat_gex.shape, mat_gex_filtered.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a477f67",
   "metadata": {},
   "source": [
    "## Cell type label hierarchy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "0519f274",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nonProlif_Proliferating</th>\n",
       "      <th>nonNK_NK</th>\n",
       "      <th>CD4CD8_Treg_MAIT</th>\n",
       "      <th>Naive_Eff_Memory</th>\n",
       "      <th>CD4Naive_CD8Naive</th>\n",
       "      <th>CD4TEM_CD8TEM_CD4CTL</th>\n",
       "      <th>CD4TCM_CD8TCM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4 CTL</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4 Naive</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4 Proliferating</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4 TCM_1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4 TCM_2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   nonProlif_Proliferating  nonNK_NK  CD4CD8_Treg_MAIT  \\\n",
       "celltype                                                                 \n",
       "CD4 CTL                                0.0       0.0               0.0   \n",
       "CD4 Naive                              0.0       0.0               0.0   \n",
       "CD4 Proliferating                      1.0       NaN               NaN   \n",
       "CD4 TCM_1                              0.0       0.0               0.0   \n",
       "CD4 TCM_2                              0.0       0.0               0.0   \n",
       "\n",
       "                   Naive_Eff_Memory  CD4Naive_CD8Naive  CD4TEM_CD8TEM_CD4CTL  \\\n",
       "celltype                                                                       \n",
       "CD4 CTL                         1.0                NaN                   2.0   \n",
       "CD4 Naive                       0.0                0.0                   NaN   \n",
       "CD4 Proliferating               NaN                NaN                   NaN   \n",
       "CD4 TCM_1                       2.0                NaN                   NaN   \n",
       "CD4 TCM_2                       2.0                NaN                   NaN   \n",
       "\n",
       "                   CD4TCM_CD8TCM  \n",
       "celltype                          \n",
       "CD4 CTL                      NaN  \n",
       "CD4 Naive                    NaN  \n",
       "CD4 Proliferating            NaN  \n",
       "CD4 TCM_1                    0.0  \n",
       "CD4 TCM_2                    0.0  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hierarchy = pd.read_csv(\"data/celltypes_hierarchy.csv\", index_col=\"celltype\")\n",
    "df_hierarchy.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "ddd8a175",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CelltypeDefException(Exception):\n",
    "    pass\n",
    "\n",
    "\n",
    "class CellTypeClassifier:\n",
    "    \n",
    "    def __init__(self, task, labels, members):\n",
    "        \n",
    "        self.task = task\n",
    "        self.labels = labels\n",
    "        self.members, self.members_flat = self.add_members(labels, members)\n",
    "        self.from_label = self.get_from_labels()\n",
    "        \n",
    "    \n",
    "    @staticmethod\n",
    "    def add_members(labels, member_dict):\n",
    "        \n",
    "        members = defaultdict(set)\n",
    "        members_flat = set()\n",
    "        for member, assign in member_dict.items():\n",
    "            if not pd.isna(assign):\n",
    "                members[labels[int(assign)]].add(member)\n",
    "                members_flat.add(member)\n",
    "\n",
    "        return members, members_flat\n",
    "\n",
    "    \n",
    "    def get_members(self, label):\n",
    "        return self.members[label]\n",
    "    \n",
    "    \n",
    "    def get_from_labels(self):\n",
    "        \n",
    "        lookup = dict()\n",
    "        for label, members in self.members.items():\n",
    "            for member in members:\n",
    "                lookup[member] = label\n",
    "        \n",
    "        return lookup\n",
    "    \n",
    "    def lookup_label(self, member):\n",
    "        return self.from_label[member]\n",
    "                  \n",
    "\n",
    "\n",
    "class CellTypeHierarchy:\n",
    "    \n",
    "    def __init__(self, hierarchy_csv):\n",
    "        self.src_table = pd.read_csv(hierarchy_csv, index_col=\"celltype\")\n",
    "        \n",
    "        self.hierachy = OrderedDict()\n",
    "        for task, members in self.src_table.to_dict().items():\n",
    "            labels = task.split(\"_\")\n",
    "            self.hierachy[task] = CellTypeClassifier(task, labels, members) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "126bbe60",
   "metadata": {},
   "outputs": [],
   "source": [
    "cth = CellTypeHierarchy(\"data/celltypes_hierarchy.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "94ff672b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "odict_keys(['nonProlif_Proliferating', 'nonNK_NK', 'CD4CD8_Treg_MAIT', 'Naive_Eff_Memory', 'CD4Naive_CD8Naive', 'CD4TEM_CD8TEM_CD4CTL', 'CD4TCM_CD8TCM', 'CD8TermEff_CD8EffMem', 'CD8TCM1_CD8TCM2_CD8TCM3', 'CD8TEM2_CD8TEM4_CD8TEM5', 'CD8TEM1_CD8TEM3_CD8TEM6'])"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cth.hierachy.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b1e42157",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Naive', 'Eff', 'Memory']"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cth.hierachy[\"Naive_Eff_Memory\"].labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "id": "6e100d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd_hierachy = pd.concat([pd.DataFrame({\"celltype\": ct.labels, \"task\": task}) for task, ct in cth.hierachy.items()], axis=0)\n",
    "pd_hierachy.to_csv(\"results/cls_task_labels.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f612c2b7",
   "metadata": {},
   "source": [
    "## Prepare matrix for fitting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "fb6cda24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(Counter({'CD8TEM': 11727, 'CD4CTL': 1736, 'CD4TEM': 4282}), (17745, 488))"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "task = \"CD4TEM_CD8TEM_CD4CTL\"\n",
    "mat_gex_filtered = mat_gex[mat_gex.obs[\"celltype.l3\"].isin(cth.hierachy[task].members_flat), selected_genes]\n",
    "\n",
    "labels = cth.hierachy[task].labels\n",
    "\n",
    "all_labels = mat_gex_filtered.obs[\"celltype.l3\"].apply(cth.hierachy[task].lookup_label)\n",
    "Counter(all_labels), mat_gex_filtered.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9427f932",
   "metadata": {},
   "source": [
    "#### Sample each class to avoid over-dominance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "b5107372",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_banlanced_labels(all_labels):\n",
    "    count = Counter(all_labels)\n",
    "    count_min = min(count.values())\n",
    "    \n",
    "    sampled = []\n",
    "    for label in labels:        \n",
    "        df_sample = mat_gex_filtered.obs.loc[all_labels == label, :]\n",
    "        \n",
    "        if count[label] > 5*count_min:\n",
    "            rate = 5*count_min/count[label]\n",
    "            \n",
    "            df_sample = df_sample.groupby(\"celltype.l3\").apply(lambda x: x.sample(frac=rate))\n",
    "            df_sample.index = df_sample.index.droplevel(\"celltype.l3\")\n",
    "            \n",
    "        sampled.append(df_sample)\n",
    "\n",
    "    sampled_bcs = pd.concat(sampled, axis=0).index\n",
    "    return sampled_bcs\n",
    "\n",
    "def rebanlance_data(data_mat, labels):\n",
    "    \n",
    "    banlanced_labels = get_banlanced_labels(labels)\n",
    "    return data_mat[banlanced_labels, :], labels[banlanced_labels]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "1d8fd1c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17745,)"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_labels.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "9c676622",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((14698, 488), (14698,))"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mat_gex_balanced, balanced_labels = rebanlance_data(mat_gex_filtered, all_labels)\n",
    "mat_gex_balanced.shape, balanced_labels.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "3cb541b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'CD4TEM': 4282, 'CD8TEM': 8680, 'CD4CTL': 1736})"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(balanced_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ad04127",
   "metadata": {},
   "source": [
    "## Train classifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "0a282d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_model_nonzero_coef(model, feature_names):\n",
    "    coef = model.coef_.T\n",
    "    nonzero = coef.sum(axis=1) > 0\n",
    "    classes = model.classes_ if coef.shape[1] > 1 else model.classes_[0]\n",
    "    \n",
    "    return  pd.DataFrame(coef[nonzero], columns=classes, index=feature_names[nonzero])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "346bf4a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((38084, 488), (38084,), (25390, 488), (25390,))"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Multi-class\n",
    "X = mat_gex_filtered.X\n",
    "y = all_labels\n",
    "\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0)\n",
    "\n",
    "X_train.shape, y_train.shape, X_test.shape, y_test.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "aa74bf3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 9min 27s, sys: 8min, total: 17min 28s\n",
      "Wall time: 4min 24s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "LogitNet(lambda_path=array([1.00000000e-01, 8.28642773e-02, 6.86648845e-02, 5.68986603e-02,\n",
       "       4.71486636e-02, 3.90693994e-02, 3.23745754e-02, 2.68269580e-02,\n",
       "       2.22299648e-02, 1.84206997e-02, 1.52641797e-02, 1.26485522e-02,\n",
       "       1.04811313e-02, 8.68511374e-03, 7.19685673e-03, 5.96362332e-03,\n",
       "       4.94171336e-03, 4.09491506e-03, 3.39322177e-03, 2.81176870e-03,\n",
       "       2.32995181e-03, 1.93069...\n",
       "       5.17947468e-04, 4.29193426e-04, 3.55648031e-04, 2.94705170e-04,\n",
       "       2.44205309e-04, 2.02358965e-04, 1.67683294e-04, 1.38949549e-04,\n",
       "       1.15139540e-04, 9.54095476e-05, 7.90604321e-05, 6.55128557e-05,\n",
       "       5.42867544e-05, 4.49843267e-05, 3.72759372e-05, 3.08884360e-05,\n",
       "       2.55954792e-05, 2.12095089e-05, 1.75751062e-05, 1.45634848e-05,\n",
       "       1.20679264e-05, 1.00000000e-05]),\n",
       "         n_jobs=10, random_state=0, standardize=False)"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# Fitting (1min 25s)\n",
    "lambda_path = np.power(10, -np.linspace(1,5,50))\n",
    "model = LogitNet(alpha=1, standardize=False, random_state=0, lambda_path=lambda_path, n_jobs=10)\n",
    "model.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "f0fc804b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Eff</th>\n",
       "      <th>Memory</th>\n",
       "      <th>Naive</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>0.045319</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABHD17A</th>\n",
       "      <td>0.020972</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC243960.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.048518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTN1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.088092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADGRE5</th>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VIM</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.084371</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XCL1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001574</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFP36L2</th>\n",
       "      <td>0.068557</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.005143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF683</th>\n",
       "      <td>0.103623</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF831</th>\n",
       "      <td>0.006599</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>159 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Eff    Memory     Naive\n",
       "A2M-AS1     0.045319  0.000000  0.000000\n",
       "ABHD17A     0.020972  0.000000  0.000000\n",
       "AC243960.1  0.000000  0.000000  0.048518\n",
       "ACTN1       0.000000  0.000000  0.088092\n",
       "ADGRE5      0.002184  0.000000  0.000000\n",
       "...              ...       ...       ...\n",
       "VIM         0.000000  0.084371  0.000000\n",
       "XCL1        0.000000  0.001574  0.000000\n",
       "ZFP36L2     0.068557  0.000000 -0.005143\n",
       "ZNF683      0.103623  0.000000  0.000000\n",
       "ZNF831      0.006599  0.000000  0.000000\n",
       "\n",
       "[159 rows x 3 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_feat_logit = get_model_nonzero_coef(model, mat_gex_filtered.var_names)\n",
    "df_feat_logit"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e23646a3",
   "metadata": {},
   "source": [
    "# Run all classifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 628,
   "id": "bf605eb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_time_now():\n",
    "    return datetime.now().strftime(\"%H:%M:%S\")\n",
    "\n",
    "def get_model_coef(model, feature_names):\n",
    "    \n",
    "    if len(model.classes_) > 2:\n",
    "        df_coef = pd.DataFrame(model.coef_.T, columns=model.classes_)\n",
    "\n",
    "    else:\n",
    "        df_coef = pd.DataFrame({model.classes_[0]: -1*model.coef_[0]})\n",
    "    \n",
    "    df_coef[\"gene_name\"] = feature_names\n",
    "    df_coef[\"task\"] = task\n",
    "    df_coef = df_coef.melt(id_vars = [\"gene_name\", \"task\"], var_name=\"celltype\", value_name=\"coeff\")\n",
    "\n",
    "    return df_coef\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfbf0e0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[14:35:27] Building classifier for nonProlif_Proliferating...\n",
      "\t[14:35:27] Rebalancing data...\n",
      "\t[14:35:28] Preparing traning and testing data...\n",
      "\t[14:35:28] Fitting the model...\n",
      "\t[14:35:29] nonProlif_Proliferating: 0.9925650557620818\n",
      "\t[14:35:29] Done!\n",
      "[14:35:29] Building classifier for nonNK_NK...\n",
      "\t[14:35:29] Rebalancing data...\n",
      "\t[14:35:29] Preparing traning and testing data...\n",
      "\t[14:35:32] Fitting the model...\n",
      "\t[14:37:11] nonNK_NK: 0.9985037406483791\n",
      "\t[14:37:11] Done!\n",
      "[14:37:11] Building classifier for CD4CD8_Treg_MAIT...\n",
      "\t[14:37:11] Rebalancing data...\n",
      "\t[14:37:11] Preparing traning and testing data...\n",
      "\t[14:37:11] Fitting the model...\n",
      "\t[14:40:01] CD4CD8_Treg_MAIT: 0.9725182277061133\n",
      "\t[14:40:01] Done!\n",
      "[14:40:01] Building classifier for Naive_Eff_Memory...\n",
      "\t[14:40:01] Rebalancing data...\n",
      "\t[14:40:01] Preparing traning and testing data...\n",
      "\t[14:40:03] Fitting the model...\n",
      "\t[14:44:02] Naive_Eff_Memory: 0.9473297274589088\n",
      "\t[14:44:02] Done!\n",
      "[14:44:02] Building classifier for CD4Naive_CD8Naive...\n",
      "\t[14:44:02] Rebalancing data...\n",
      "\t[14:44:02] Preparing traning and testing data...\n",
      "\t[14:44:03] Fitting the model...\n",
      "\t[14:44:23] CD4Naive_CD8Naive: 0.9910586552217453\n",
      "\t[14:44:23] Done!\n",
      "[14:44:23] Building classifier for CD4TEM_CD8TEM_CD4CTL...\n",
      "\t[14:44:23] Rebalancing data...\n",
      "\t[14:44:23] Preparing traning and testing data...\n",
      "\t[14:44:23] Fitting the model...\n",
      "\t[14:45:13] CD4TEM_CD8TEM_CD4CTL: 0.9816326530612245\n",
      "\t[14:45:13] Done!\n",
      "[14:45:13] Building classifier for CD4TCM_CD8TCM...\n",
      "\t[14:45:13] Rebalancing data...\n",
      "\t[14:45:14] Preparing traning and testing data...\n",
      "\t[14:45:14] Fitting the model...\n",
      "\t[14:46:40] CD4TCM_CD8TCM: 0.981888246628131\n",
      "\t[14:46:40] Done!\n",
      "[14:46:40] Building classifier for CD8TermEff_CD8EffMem...\n",
      "\t[14:46:40] Rebalancing data...\n",
      "\t[14:46:40] Preparing traning and testing data...\n",
      "\t[14:46:40] Fitting the model...\n",
      "\t[14:47:04] CD8TermEff_CD8EffMem: 0.9738562091503268\n",
      "\t[14:47:04] Done!\n",
      "[14:47:04] Building classifier for CD8TCM1_CD8TCM2_CD8TCM3...\n",
      "\t[14:47:04] Rebalancing data...\n",
      "\t[14:47:04] Preparing traning and testing data...\n",
      "\t[14:47:04] Fitting the model...\n",
      "\t[14:47:19] CD8TCM1_CD8TCM2_CD8TCM3: 0.8589595375722543\n",
      "\t[14:47:19] Done!\n",
      "[14:47:19] Building classifier for CD8TEM2_CD8TEM4_CD8TEM5...\n",
      "\t[14:47:19] Rebalancing data...\n",
      "\t[14:47:19] Preparing traning and testing data...\n",
      "\t[14:47:19] Fitting the model...\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "models = {}\n",
    "df_coefs = []\n",
    "for task, ct_obj in cth.hierachy.items():\n",
    "    \n",
    "    result = {}\n",
    "    \n",
    "    print(\"[{}] Building classifier for {}...\".format(get_time_now(), task))\n",
    "    mat_gex_filtered = mat_gex[mat_gex.obs[\"celltype.l3\"].isin(cth.hierachy[task].members_flat), selected_genes]\n",
    "    labels = cth.hierachy[task].labels\n",
    "    all_labels = mat_gex_filtered.obs[\"celltype.l3\"].apply(cth.hierachy[task].lookup_label)\n",
    "\n",
    "    print(\"\\t[{}] Rebalancing data...\".format(get_time_now()))\n",
    "    mat_gex_balanced, balanced_labels = rebanlance_data(mat_gex_filtered, all_labels)\n",
    "    \n",
    "    print(\"\\t[{}] Preparing traning and testing data...\".format(get_time_now()))\n",
    "    X_train, X_test, y_train, y_test = train_test_split(mat_gex_balanced.X, \n",
    "                                                        balanced_labels, \n",
    "                                                        test_size=0.3, \n",
    "                                                        random_state=0, \n",
    "                                                        stratify=balanced_labels)\n",
    "    \n",
    "    print(\"\\t[{}] Fitting the model...\".format(get_time_now()))\n",
    "    lambda_path = np.power(10, -np.linspace(1,5,50))\n",
    "    model = LogitNet(alpha=1, standardize=False, random_state=0, lambda_path=lambda_path, n_jobs=10)\n",
    "    model.fit(X_train, y_train)\n",
    "    \n",
    "    accu =  model.score(X_test, y_test)\n",
    "    print(\"\\t[{}] {}: {}\".format(get_time_now(), task, accu))\n",
    "    \n",
    "    print(\"\\t[{}] Extracting feature coeff...\".format(get_time_now()))\n",
    "    df_coef = get_model_coef(model, selected_genes)\n",
    "    df_coefs.append(df_coef)\n",
    "        \n",
    "    print(\"\\t[{}] Done!\".format(get_time_now()))\n",
    "    \n",
    "    result = {\n",
    "        \"model\": model, \"accuracy\": accu, \"label_count\": Counter(balanced_labels), \n",
    "        #\"feat_coeff\": df_feat_logit\n",
    "    }\n",
    "    models[task] = result\n",
    "\n",
    "df_coefs = pd.concat(df_coefs, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "6cdae147",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9926\tnonProlif_Proliferating\tCounter({'nonProlif': 3735, 'Proliferating': 747})\n",
      "0.9985\tnonNK_NK\tCounter({'nonNK': 68765, 'NK': 18116})\n",
      "0.9725\tCD4CD8_Treg_MAIT\tCounter({'CD4CD8': 12537, 'MAIT': 2784, 'Treg': 2507})\n",
      "0.9473\tNaive_Eff_Memory\tCounter({'Naive': 27957, 'Memory': 17772, 'Eff': 17745})\n",
      "0.9911\tCD4Naive_CD8Naive\tCounter({'CD4Naive': 17479, 'CD8Naive': 10478})\n",
      "0.9816\tCD4TEM_CD8TEM_CD4CTL\tCounter({'CD8TEM': 8680, 'CD4TEM': 4282, 'CD4CTL': 1736})\n",
      "0.9819\tCD4TCM_CD8TCM\tCounter({'CD4TCM': 14415, 'CD8TCM': 2883})\n",
      "0.9739\tCD8TermEff_CD8EffMem\tCounter({'CD8TermEff': 7912, 'CD8EffMem': 3815})\n",
      "0.859\tCD8TCM1_CD8TCM2_CD8TCM3\tCounter({'CD8TCM2': 1322, 'CD8TCM1': 929, 'CD8TCM3': 632})\n",
      "0.8707\tCD8TEM2_CD8TEM4_CD8TEM5\tCounter({'CD8TEM4': 3504, 'CD8TEM2': 2435, 'CD8TEM5': 1973})\n",
      "0.9588\tCD8TEM1_CD8TEM3_CD8TEM6\tCounter({'CD8TEM1': 1965, 'CD8TEM6': 636, 'CD8TEM3': 393})\n"
     ]
    }
   ],
   "source": [
    "for model, content in models.items():\n",
    "    print(\"{}\\t{}\\t{}\".format(np.round(content[\"accuracy\"], 4), model, content[\"label_count\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d37083fb",
   "metadata": {},
   "source": [
    "## Save the classifier and feature coeff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 626,
   "id": "d2d8fe40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>task</th>\n",
       "      <th>celltype</th>\n",
       "      <th>coeff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>CLSPN</td>\n",
       "      <td>nonProlif_Proliferating</td>\n",
       "      <td>Proliferating</td>\n",
       "      <td>0.219409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>MCM4</td>\n",
       "      <td>nonProlif_Proliferating</td>\n",
       "      <td>Proliferating</td>\n",
       "      <td>0.246096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>MKI67</td>\n",
       "      <td>nonProlif_Proliferating</td>\n",
       "      <td>Proliferating</td>\n",
       "      <td>0.249142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>PCNA</td>\n",
       "      <td>nonProlif_Proliferating</td>\n",
       "      <td>Proliferating</td>\n",
       "      <td>0.210181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>STMN1</td>\n",
       "      <td>nonProlif_Proliferating</td>\n",
       "      <td>Proliferating</td>\n",
       "      <td>0.695056</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name                     task       celltype     coeff\n",
       "102     CLSPN  nonProlif_Proliferating  Proliferating  0.219409\n",
       "275      MCM4  nonProlif_Proliferating  Proliferating  0.246096\n",
       "279     MKI67  nonProlif_Proliferating  Proliferating  0.249142\n",
       "330      PCNA  nonProlif_Proliferating  Proliferating  0.210181\n",
       "433     STMN1  nonProlif_Proliferating  Proliferating  0.695056"
      ]
     },
     "execution_count": 626,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_coefs.query(\"celltype == 'Proliferating' & coeff > 0.2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "551e18de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the classifier\n",
    "with open('models/hc_allmodels_seurat_genes.pkl', 'wb') as fid:\n",
    "    pickle.dump(models, fid)    \n",
    "    \n",
    "# save model coeff\n",
    "df_coefs.to_csv(\"models/featcoef_allmodels_seurat_genes.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2337412a",
   "metadata": {},
   "source": [
    "## Prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "051df205",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(60728, 84)"
      ]
     },
     "execution_count": 281,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mat_pact476 = pd.read_csv(\"data/PACT476_filtered_rsem_matrix.csv\", index_col=0)\n",
    "df_mat_pact476.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "id": "71cebf90",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_pact476 = np.array(df_mat_pact476.loc[common_genes, :]).T\n",
    "df_obs_pact476 = pd.DataFrame({\"patient_id\": \"PACT476\"}, index=df_mat_pact476.columns)\n",
    "df_var_pact476 = pd.DataFrame(index=common_genes)\n",
    "\n",
    "adata_pact476 = sc.AnnData(mat_pact476, obs=df_obs_pact476, var=df_var_pact476)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "id": "6ba56db2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((84, 20177), (84, 488))"
      ]
     },
     "execution_count": 312,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc.pp.normalize_total(adata_pact476, target_sum=10000)\n",
    "sc.pp.log1p(adata_pact476)\n",
    "sc.pp.scale(adata_pact476, max_value=10)\n",
    "\n",
    "adata_pact476_filtered = adata_pact476[:, selected_genes]\n",
    "adata_pact476.shape, adata_pact476_filtered.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "id": "d36d3be0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 75 ms, sys: 281 ms, total: 356 ms\n",
      "Wall time: 18.7 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Proliferating</th>\n",
       "      <th>nonProlif</th>\n",
       "      <th>NK</th>\n",
       "      <th>nonNK</th>\n",
       "      <th>CD4CD8</th>\n",
       "      <th>MAIT</th>\n",
       "      <th>Treg</th>\n",
       "      <th>Eff</th>\n",
       "      <th>Memory</th>\n",
       "      <th>Naive</th>\n",
       "      <th>...</th>\n",
       "      <th>CD8TermEff</th>\n",
       "      <th>CD8TCM1</th>\n",
       "      <th>CD8TCM2</th>\n",
       "      <th>CD8TCM3</th>\n",
       "      <th>CD8TEM2</th>\n",
       "      <th>CD8TEM4</th>\n",
       "      <th>CD8TEM5</th>\n",
       "      <th>CD8TEM1</th>\n",
       "      <th>CD8TEM3</th>\n",
       "      <th>CD8TEM6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCR24_T02.P02-C05</th>\n",
       "      <td>0.002261</td>\n",
       "      <td>0.997739</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.999964</td>\n",
       "      <td>0.882477</td>\n",
       "      <td>0.065418</td>\n",
       "      <td>0.052105</td>\n",
       "      <td>0.521385</td>\n",
       "      <td>0.450793</td>\n",
       "      <td>0.027822</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009165</td>\n",
       "      <td>0.734952</td>\n",
       "      <td>0.101459</td>\n",
       "      <td>0.163589</td>\n",
       "      <td>0.653255</td>\n",
       "      <td>0.297265</td>\n",
       "      <td>0.049480</td>\n",
       "      <td>0.751994</td>\n",
       "      <td>0.096663</td>\n",
       "      <td>0.151342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCR7_T01.P01-B08</th>\n",
       "      <td>0.001700</td>\n",
       "      <td>0.998300</td>\n",
       "      <td>0.106468</td>\n",
       "      <td>0.893532</td>\n",
       "      <td>0.537771</td>\n",
       "      <td>0.009957</td>\n",
       "      <td>0.452273</td>\n",
       "      <td>0.040037</td>\n",
       "      <td>0.098499</td>\n",
       "      <td>0.861464</td>\n",
       "      <td>...</td>\n",
       "      <td>0.164089</td>\n",
       "      <td>0.252467</td>\n",
       "      <td>0.419640</td>\n",
       "      <td>0.327893</td>\n",
       "      <td>0.634037</td>\n",
       "      <td>0.310440</td>\n",
       "      <td>0.055523</td>\n",
       "      <td>0.373041</td>\n",
       "      <td>0.080416</td>\n",
       "      <td>0.546544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCR14_T02.P02-H10</th>\n",
       "      <td>0.001670</td>\n",
       "      <td>0.998330</td>\n",
       "      <td>0.001984</td>\n",
       "      <td>0.998016</td>\n",
       "      <td>0.672041</td>\n",
       "      <td>0.289703</td>\n",
       "      <td>0.038256</td>\n",
       "      <td>0.343269</td>\n",
       "      <td>0.346660</td>\n",
       "      <td>0.310072</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024515</td>\n",
       "      <td>0.431873</td>\n",
       "      <td>0.429938</td>\n",
       "      <td>0.138190</td>\n",
       "      <td>0.633261</td>\n",
       "      <td>0.329728</td>\n",
       "      <td>0.037011</td>\n",
       "      <td>0.675094</td>\n",
       "      <td>0.140698</td>\n",
       "      <td>0.184209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCR19_T02.P02-C04</th>\n",
       "      <td>0.002287</td>\n",
       "      <td>0.997713</td>\n",
       "      <td>0.000318</td>\n",
       "      <td>0.999682</td>\n",
       "      <td>0.894146</td>\n",
       "      <td>0.036383</td>\n",
       "      <td>0.069471</td>\n",
       "      <td>0.093960</td>\n",
       "      <td>0.816659</td>\n",
       "      <td>0.089381</td>\n",
       "      <td>...</td>\n",
       "      <td>0.183001</td>\n",
       "      <td>0.350793</td>\n",
       "      <td>0.559307</td>\n",
       "      <td>0.089900</td>\n",
       "      <td>0.618150</td>\n",
       "      <td>0.303279</td>\n",
       "      <td>0.078571</td>\n",
       "      <td>0.418601</td>\n",
       "      <td>0.093419</td>\n",
       "      <td>0.487980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCR5_T02.P04-F08</th>\n",
       "      <td>0.020707</td>\n",
       "      <td>0.979293</td>\n",
       "      <td>0.002257</td>\n",
       "      <td>0.997743</td>\n",
       "      <td>0.736118</td>\n",
       "      <td>0.036370</td>\n",
       "      <td>0.227512</td>\n",
       "      <td>0.318503</td>\n",
       "      <td>0.651296</td>\n",
       "      <td>0.030202</td>\n",
       "      <td>...</td>\n",
       "      <td>0.549194</td>\n",
       "      <td>0.337891</td>\n",
       "      <td>0.508925</td>\n",
       "      <td>0.153183</td>\n",
       "      <td>0.386772</td>\n",
       "      <td>0.396742</td>\n",
       "      <td>0.216486</td>\n",
       "      <td>0.246987</td>\n",
       "      <td>0.354985</td>\n",
       "      <td>0.398027</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Proliferating  nonProlif        NK     nonNK    CD4CD8  \\\n",
       "TCR24_T02.P02-C05       0.002261   0.997739  0.000036  0.999964  0.882477   \n",
       "TCR7_T01.P01-B08        0.001700   0.998300  0.106468  0.893532  0.537771   \n",
       "TCR14_T02.P02-H10       0.001670   0.998330  0.001984  0.998016  0.672041   \n",
       "TCR19_T02.P02-C04       0.002287   0.997713  0.000318  0.999682  0.894146   \n",
       "TCR5_T02.P04-F08        0.020707   0.979293  0.002257  0.997743  0.736118   \n",
       "\n",
       "                       MAIT      Treg       Eff    Memory     Naive  ...  \\\n",
       "TCR24_T02.P02-C05  0.065418  0.052105  0.521385  0.450793  0.027822  ...   \n",
       "TCR7_T01.P01-B08   0.009957  0.452273  0.040037  0.098499  0.861464  ...   \n",
       "TCR14_T02.P02-H10  0.289703  0.038256  0.343269  0.346660  0.310072  ...   \n",
       "TCR19_T02.P02-C04  0.036383  0.069471  0.093960  0.816659  0.089381  ...   \n",
       "TCR5_T02.P04-F08   0.036370  0.227512  0.318503  0.651296  0.030202  ...   \n",
       "\n",
       "                   CD8TermEff   CD8TCM1   CD8TCM2   CD8TCM3   CD8TEM2  \\\n",
       "TCR24_T02.P02-C05    0.009165  0.734952  0.101459  0.163589  0.653255   \n",
       "TCR7_T01.P01-B08     0.164089  0.252467  0.419640  0.327893  0.634037   \n",
       "TCR14_T02.P02-H10    0.024515  0.431873  0.429938  0.138190  0.633261   \n",
       "TCR19_T02.P02-C04    0.183001  0.350793  0.559307  0.089900  0.618150   \n",
       "TCR5_T02.P04-F08     0.549194  0.337891  0.508925  0.153183  0.386772   \n",
       "\n",
       "                    CD8TEM4   CD8TEM5   CD8TEM1   CD8TEM3   CD8TEM6  \n",
       "TCR24_T02.P02-C05  0.297265  0.049480  0.751994  0.096663  0.151342  \n",
       "TCR7_T01.P01-B08   0.310440  0.055523  0.373041  0.080416  0.546544  \n",
       "TCR14_T02.P02-H10  0.329728  0.037011  0.675094  0.140698  0.184209  \n",
       "TCR19_T02.P02-C04  0.303279  0.078571  0.418601  0.093419  0.487980  \n",
       "TCR5_T02.P04-F08   0.396742  0.216486  0.246987  0.354985  0.398027  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 340,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "df_prob = []\n",
    "for task, model_obj in models.items():\n",
    "    \n",
    "    prob = pd.DataFrame(model_obj[\"model\"].predict_proba(adata_pact476_filtered.X), \n",
    "                        columns=model_obj[\"model\"].classes_, \n",
    "                        index=adata_pact476_filtered.obs_names)\n",
    "    df_prob.append(prob)\n",
    "\n",
    "df_prob = pd.concat(df_prob, axis=1)\n",
    "df_prob.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "id": "f4dac9d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_prob.to_csv(\"results/PACT476_predprob_seurat_genes.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "770c6bdf",
   "metadata": {},
   "source": [
    "## Output decision tree path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 450,
   "id": "17eb68ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "DECISION_TREE = {\n",
    "    \"Proliferating\": {},\n",
    "    \"nonProlif\": {\"NK\", \"nonNK\"},\n",
    "    \"NK\": {},\n",
    "    \"nonNK\": {\"CD4CD8\", \"MAIT\", \"Treg\"},\n",
    "    \"CD4CD8\": {\"Eff\", \"Memory\", \"Naive\"},\n",
    "    \"MAIT\": {},\n",
    "    \"Treg\": {},\n",
    "    \"Eff\": {\"CD4CTL\", \"CD4TEM\", \"CD8TEM\"},\n",
    "    \"Memory\": {\"CD4TCM\", \"CD8TCM\"},\n",
    "    \"Naive\": {\"CD4Naive\", \"CD8Naive\"},\n",
    "    \"CD4Naive\": {},\n",
    "    \"CD8Naive\": {},\n",
    "    \"CD4CTL\": {},\n",
    "    \"CD4TEM\": {},\n",
    "    \"CD8TEM\": {\"CD8EffMem\", \"CD8TermEff\"},\n",
    "    \"CD4TCM\": {},\n",
    "    \"CD8TCM\": {\"CD8TCM1\", \"CD8TCM2\", \"CD8TCM3\"},\n",
    "    \"CD8EffMem\": {\"CD8TEM1\", \"CD8TEM3\", \"CD8TEM6\"},\n",
    "    \"CD8TermEff\": {\"CD8TEM2\", \"CD8TEM4\", \"CD8TEM5\"},\n",
    "    \"CD8TCM1\": {},\n",
    "    \"CD8TCM2\": {},\n",
    "    \"CD8TCM3\": {},\n",
    "    \"CD8TEM2\": {},\n",
    "    \"CD8TEM4\": {},\n",
    "    \"CD8TEM5\": {},\n",
    "    \"CD8TEM1\": {},\n",
    "    \"CD8TEM3\": {},\n",
    "    \"CD8TEM6\": {}\n",
    "}\n",
    "TERMINAL_TYPES = set([k for k,v in DECISION_TREE.items() if not v])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 540,
   "id": "c0aef577",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_candidates(winner, candidates):\n",
    "    if DECISION_TREE[winner]:\n",
    "        for sub in DECISION_TREE[winner]:\n",
    "            get_candidates(sub, candidates)\n",
    "    else:\n",
    "        return candidates.append(winner)\n",
    "    \n",
    "    return candidates\n",
    "\n",
    "def find_path(score):\n",
    "    candidates = TERMINAL_TYPES.copy()\n",
    "    path = []\n",
    "    for task, ct in cth.hierachy.items():\n",
    "        winner, prob = score[ct.labels].idxmax(),  score[ct.labels].max()\n",
    "        winner_prob = \"{} ({})\".format(winner, np.round(prob, 4))\n",
    "        \n",
    "        avail = get_candidates(winner, [])\n",
    "        if avail:\n",
    "            refined = candidates.intersection(avail)\n",
    "            if refined:\n",
    "                path.append(winner_prob)\n",
    "                candidates = refined\n",
    "\n",
    "        if winner in candidates:\n",
    "            final = winner\n",
    "            path.append(winner_prob)\n",
    "    \n",
    "    return \"-> \".join(path), final\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 541,
   "id": "21373e81",
   "metadata": {},
   "outputs": [],
   "source": [
    "paths, finals = [], []\n",
    "for _, row in df_prob.iterrows():\n",
    "    path, final = find_path(row)\n",
    "    paths.append(path)\n",
    "    finals.append(final)\n",
    "\n",
    "df_path = pd.DataFrame({\"cell_id\":df_prob.index, \"pred\": finals, \"path\": paths})\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 545,
   "id": "1ab471bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>pred</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCR24_T02.P02-C05</td>\n",
       "      <td>CD8TEM1</td>\n",
       "      <td>nonProlif (0.9977)-&gt; nonNK (1.0)-&gt; CD4CD8 (0.8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCR7_T01.P01-B08</td>\n",
       "      <td>CD8Naive</td>\n",
       "      <td>nonProlif (0.9983)-&gt; nonNK (0.8935)-&gt; CD4CD8 (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCR14_T02.P02-H10</td>\n",
       "      <td>CD8TCM1</td>\n",
       "      <td>nonProlif (0.9983)-&gt; nonNK (0.998)-&gt; CD4CD8 (0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCR19_T02.P02-C04</td>\n",
       "      <td>CD8TCM2</td>\n",
       "      <td>nonProlif (0.9977)-&gt; nonNK (0.9997)-&gt; CD4CD8 (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCR5_T02.P04-F08</td>\n",
       "      <td>CD8TCM2</td>\n",
       "      <td>nonProlif (0.9793)-&gt; nonNK (0.9977)-&gt; CD4CD8 (...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cell_id      pred  \\\n",
       "0  TCR24_T02.P02-C05   CD8TEM1   \n",
       "1   TCR7_T01.P01-B08  CD8Naive   \n",
       "2  TCR14_T02.P02-H10   CD8TCM1   \n",
       "3  TCR19_T02.P02-C04   CD8TCM2   \n",
       "4   TCR5_T02.P04-F08   CD8TCM2   \n",
       "\n",
       "                                                path  \n",
       "0  nonProlif (0.9977)-> nonNK (1.0)-> CD4CD8 (0.8...  \n",
       "1  nonProlif (0.9983)-> nonNK (0.8935)-> CD4CD8 (...  \n",
       "2  nonProlif (0.9983)-> nonNK (0.998)-> CD4CD8 (0...  \n",
       "3  nonProlif (0.9977)-> nonNK (0.9997)-> CD4CD8 (...  \n",
       "4  nonProlif (0.9793)-> nonNK (0.9977)-> CD4CD8 (...  "
      ]
     },
     "execution_count": 545,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_path.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 544,
   "id": "4d832c4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_path.to_csv(\"results/PACT476_predpath_seurat_genes.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44ded994",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
